src/js/controller/controller.js

```diff
@@ -710,7 +710,8 @@ Object.assign(Controller.prototype, {
             _actionOnAttach = _completeHandler;
 
             const idx = _model.get('item');
-            if (idx === _model.get('playlist').length - 1) {
+            let overrideAutoplay = _model.get('__autoplayOverride');
+            if (overrideAutoplay || idx === _model.get('playlist').length - 1) {
```

Comment:
i wanted to use a different conditional here to be cleaner which would base whether or not to stop on the presence of a `nextUp` item however this does not cover the open source case where `nextUp` never gets set on the model (no recs). in that case you would still need the playlist check but could check `nextUp` in the event recs is present. super messy. this keeps default OS behavior while giving the application a chance to override this behavior if we want to do something smarter.

I tried this out with a extension method in commercial but since the entirety of this function would have been repeated (or a flag used as a parameter) it didn't seem like the right use case for it. along with that, the extended `onComplete` method in commercial would have had to know recs state at any point in time to make the right determination which seems like way more of the plugin's responsibility. 

This allows the _plugin_ to have control over the autoplay override and stops unnecessary code duplication or parameterization in the commercial controller

Reply:
I think it would be better to break this method up into the components you need to override in commercial.

```
this.completeHandler = function() {
   if (this.completeCancelled()) { // encapsulate _isIdle & _stopPlaylist check
      return;
  }
  _actionOnAttach = _completeHandler;
  const idx = _model.get('item');
  const lastItemInPlaylist = idx === _model.get('playlist').length - 1;
  if (lastItemInPlaylist) { // this could be a method too if you want to add a nextUp check
    this.lastItemComplete();
    return;
  }
  // Aren't we just trying to interrupt this method?
  _this.nextItem();
}
```

Reply:
I'm not a fan of communicating between modules using the model (side-effects). The above would allow for a more functional approach.